jvb/src/main/kotlin/org/jitsi/videobridge/cc/allocation/BandwidthAllocator.kt

```diff
@@ -0,0 +1,397 @@
+     *
+     * @param conferenceMediaSources the list of endpoint media sources in order of priority to allocate for.
+     * @return the new [BandwidthAllocation].
+     */
+    @Synchronized
+    private fun allocate(conferenceMediaSources: List<MediaSourceDesc>): BandwidthAllocation {
+        val sourceBitrateAllocations = createAllocations(conferenceMediaSources)
+        if (sourceBitrateAllocations.isEmpty()) {
+            return BandwidthAllocation(emptySet())
+        }
+        var remainingBandwidth = availableBandwidth
+        var oldRemainingBandwidth: Long = -1
+        var oversending = false
+        while (oldRemainingBandwidth != remainingBandwidth) {
+            oldRemainingBandwidth = remainingBandwidth
+            for (i in sourceBitrateAllocations.indices) {
+                val sourceBitrateAllocation = sourceBitrateAllocations[i]
+                if (sourceBitrateAllocation.constraints.isDisabled()) {
+                    continue
+                }
+
+                // In stage view improve greedily until preferred, in tile view go step-by-step.
+                remainingBandwidth -= sourceBitrateAllocation.improve(remainingBandwidth, i == 0)
+                if (remainingBandwidth < 0) {
+                    oversending = true
+                }
+
+                // In stage view, do not allocate bandwidth for thumbnails until the on-stage reaches "preferred".
+                // This prevents enabling thumbnail only to disable them when bwe slightly increases allowing on-stage
+                // to take more.
+                if (sourceBitrateAllocation.isOnStage() && !sourceBitrateAllocation.hasReachedPreferred()) {
+                    break
+                }
+            }
+        }
+
+        // The sources which are in lastN, and are sending video, but were suspended due to bwe.
+        val suspendedIds = sourceBitrateAllocations
+            .filter { it.isSuspended }
+            .mapNotNull { it.mediaSource.sourceName }.toList()
+        if (suspendedIds.isNotEmpty()) {
+            logger.info("Sources suspended due to insufficient bandwidth (bwe=$availableBandwidth bps): $suspendedIds")
+        }
+        val allocations: MutableSet<SingleAllocation> = HashSet()
+        var targetBps: Long = 0
+        var idealBps: Long = 0
+        for (sourceBitrateAllocation: SingleSourceAllocation in sourceBitrateAllocations) {
+            allocations.add(sourceBitrateAllocation.result)
+            targetBps += sourceBitrateAllocation.targetBitrate
+            idealBps += sourceBitrateAllocation.idealBitrate
+        }
+        return BandwidthAllocation(allocations, oversending, idealBps, targetBps, suspendedIds)
+    }
+
+    /**
+     * Query whether this allocator is forwarding a source from a given endpoint, as of its
+     * most recent allocation decision.
+     */
+    fun isForwarding(endpointId: String): Boolean {
+        return allocation.isForwarding(endpointId)
+    }
```

Comment:
I might use Kotlin `=` syntax for a one-liner like this.

Reply:
I removed the function, it wasn't used